contracts/solidity/NFTXYieldStakingZap.sol

```diff
@@ -0,0 +1,336 @@
+    uint256 vaultId,
+    bytes calldata swapCallData
+  ) external payable nonReentrant {
+    // Ensure we have tx value
+    require(msg.value > 0, 'Invalid value provided');
+
+    // Get our start WETH balance
+    uint wethBalance = WETH.balanceOf(address(this));
+
+    // Wrap ETH into WETH for our contract from the sender
+    if (msg.value > 0) {
+      WETH.deposit{value: msg.value}();
+    }
+
+    // Get our vaults base staking token. This is used to calculate the xToken
+    address baseToken = _vaultAddress(vaultId);
+    require(baseToken != address(0), 'Invalid vault provided');
+
+    // Convert WETH to vault token
+    uint256 vaultTokenAmount = _fillQuote(baseToken, swapCallData);
+
+    // Make a direct timelock mint using the default timelock duration. This sends directly
+    // to our user, rather than via the zap, to avoid the timelock locking the tx.
+    IERC20Upgradeable(baseToken).transfer(inventoryStaking.vaultXToken(vaultId), vaultTokenAmount);
+    inventoryStaking.timelockMintFor(vaultId, vaultTokenAmount, msg.sender, 2);
+
+    // Return any left of WETH to the user as ETH
+    uint256 remainingWETH = WETH.balanceOf(address(this)) - wethBalance;
+    if (remainingWETH > 0) {
+      // Unwrap our WETH into ETH and transfer it to the recipient
+      WETH.withdraw(remainingWETH);
+      (bool success, ) = payable(msg.sender).call{value: remainingWETH}("");
+      require(success, "Unable to send unwrapped WETH");
+    }
+  }
+
+
+  /**
+   * @notice Allows the user to buy and stake tokens against a Liquidity pool. This will
+   * handle the purchase of the vault tokens against 0x, the liquidity pool supplying via
+   * sushi and then the timelocking against our LP token.
+   * 
+   * @param vaultId The ID of the NFTX vault
+   * @param swapCallData The `data` field from the 0x API response
+   * @param minTokenIn The minimum amount of token to LP
+   * @param minWethIn The minimum amount of ETH (WETH) to LP
+   * @param wethIn The amount of ETH (WETH) supplied
+   */
+
+  function buyAndStakeLiquidity(
+    // Base data
+    uint256 vaultId,
+
+    // 0x integration
+    bytes calldata swapCallData,
+
+    // Sushiswap integration
+    uint256 minTokenIn,
+    uint256 minWethIn,
+    uint256 wethIn
+
+  ) external payable nonReentrant {
+    // Ensure we have tx value
+    require(msg.value > 0, 'Invalid value provided');
+    require(msg.value > wethIn, 'Insufficient vault sent for pairing');
+
+    // Get our start WETH balance
+    uint wethBalance = WETH.balanceOf(address(this));
+
+    // Wrap ETH into WETH for our contract from the sender
+    if (msg.value > 0) {
+      WETH.deposit{value: msg.value}();
+    }
+
+    // Get our vaults base staking token. This is used to calculate the xToken
+    address baseToken = _vaultAddress(vaultId);
+    require(baseToken != address(0), 'Invalid vault provided');
+
+    // Convert WETH to vault token
+    uint256 vaultTokenAmount = _fillQuote(baseToken, swapCallData);
```

Comment:
Assuming this would be calculated to convert half of  the WETH to buy vault tokens?

Reply:
This should, in theory, allow the frontend to determine this calculation and also account for slippage based on the interaction between 0x and Sushi. Have left it flexible based on their request.